### PR TITLE
msg deduplication

### PR DESCRIPTION
**Pull request by tg123**
make life easier

code based on #22, but can be cherry-picked if you want

### CODE REVIEW

**Comment by brodyreeves** [comment]
I'll probably go ahead and grab the time filter, when I have the chance I'll deal with the merge/restructuring issues and get it pulled in. Thanks!

File: Locales/enUS.lua [hunk 1]
```diff
@@ -48,3 +52,6 @@ L["Toggle alerts on"] = "Toggle alerts on"
 L["You have no alerts"] = "You have no alerts"
 L["You have %s alert"] = "You have |cFF00FF00%s|cFFCFCFCF alert"
 L["You have %s alerts"] = "You have |cFF00FF00%s|cFFCFCFCF alerts"
+L["Deduplication"] = "Deduplication"
+L["Ignore duplicated message in"] = "Ignore duplicated message in"
+L["unit second, 0 to disable"] = "unit second, 0 to disable"
```
Comment (brodyreeves) [inline Locales/enUS.lua]: Useful for time filtering

File: Locales/enUS.lua [hunk 2]
```diff
@@ -5,6 +5,7 @@ L["Add Channel"] = "Add Channel"
 L["Add Keyword"] = "Add Keyword"
 L["Add a channel to watch from Ex: '4. LookingForGroup'"] = "Add a channel to watch from Ex: '4. LookingForGroup'"
 L["Add a keyword to watch for"] = "Add a keyword to watch for"
+L["Add a keyword to filter"] = "Add a keyword to filter"
```
Comment (brodyreeves) [inline Locales/enUS.lua]: Unnecessary due to being contained in 60d8731

File: Locales/enUS.lua [hunk 3]
```diff
@@ -20,6 +21,7 @@ L["Filter with GlobalIgnoreList"] = "Filter with GlobalIgnoreList"
 L["Ignore messages from players on your ignore list"] = "Ignore messages from players on your ignore list"
 L["Keyword"] = "Keyword"
 L["Keywords"] = "Keywords"
+L["Keywords to filter"] = "Keywords to filter"
```
Comment (brodyreeves) [inline Locales/enUS.lua]: Unnecessary due to being contained in 60d8731

File: Locales/enUS.lua [hunk 4]
```diff
@@ -35,10 +37,12 @@ L["Remove Channel"] = "Remove Channel"
 L["Remove Keyword"] = "Remove Keyword"
 L["Remove selected channel from being watched"] = "Remove selected channel from being watched"
 L["Remove selected keyword from being watched for"] = "Remove selected keyword from being watched for"
+L["Remove selected keyword from being filtered"] = "Remove selected keyword from being filtered"
```
Comment (brodyreeves) [inline Locales/enUS.lua]: Unnecessary due to being contained in 60d8731

File: Locales/enUS.lua [hunk 5]
```diff
@@ -35,10 +37,12 @@ L["Remove Channel"] = "Remove Channel"
...
 L["Right-Click"] = "Right-Click"
 L["Select a channel to remove from being watched"] = "Select a channel to remove from being watched"
 L["Select a channel to watch"] = "Select a channel to watch"
 L["Select a keyword to remove from being watched for"] = "Select a keyword to remove from being watched for"
+L["Select a keyword to remove from being filtered"] = "Select a keyword to remove from being filtered"
```
Comment (brodyreeves) [inline Locales/enUS.lua]: Unnecessary due to being contained in 60d8731

File: config.lua [hunk 6]
```diff
@@ -8,11 +8,13 @@ MyChatAlert.defaults = {
         printOn = true,
         channels = {},
         words = {},
+        filterwords = {},
```
Comment (brodyreeves) [inline config.lua]: Unnecessary due to being contained in 60d8731

File: config.lua [hunk 7]
```diff
@@ -163,9 +179,43 @@ MyChatAlert.options = {
...
+                    func = function()
+                        if filterwordsToDelete then
+                            tremove(MyChatAlert.db.profile.filterwords, filterwordsToDelete)
+                            filterwordsToDelete = nil
+                        end
+                    end,
+                    disabled = function() return not MyChatAlert.db.profile.enabled end,
+                },
+            },
+        },
```
Comment (brodyreeves) [inline config.lua]: Unnecessary due to being contained in 60d8731

File: core.lua [hunk 8]
```diff
@@ -36,17 +38,41 @@ function MyChatAlert:CHAT_MSG_CHANNEL(event, message, author, _, channel)
...
-                    if self.db.profile.soundOn then PlaySound(self.db.profile.sound) end
-                    if self.db.profile.printOn then
-                        LibStub("AceConsole-3.0"):Print(format(L["Printed alert"], word, "|Hplayer:" .. TrimRealmName(author) .. ":0|h" .. TrimRealmName(author) .. "|h", message))
+            local filtered = false
+            for k2, word in pairs(self.db.profile.filterwords) do
+                if message:lower():find(word:lower()) then
+                    filtered = true
+                    break
+                end
+            end
```
Comment (brodyreeves) [inline core.lua]: Unnecessary due to being contained in 60d8731

File: core.lua [hunk 9]
```diff
@@ -36,17 +38,41 @@ function MyChatAlert:CHAT_MSG_CHANNEL(event, message, author, _, channel)
...
+            if time() - (dedupmsg[message] or 0) < self.db.profile.dedupTime then
+                filtered = true
+            end
+
+            -- gc
+            for m, t in pairs(dedupmsg) do 
+                if time() - t > self.db.profile.dedupTime then
+                    dedupmsg[m] = nil
+                end
+            end
```
Comment (brodyreeves) [inline core.lua]: Useful for time filtering

File: config.lua [hunk 10]
```diff
@@ -8,11 +8,13 @@ MyChatAlert.defaults = {
...
         channels = {},
         words = {},
+        filterwords = {},
+        dedupTime = 0,
         globalIgnoreListFilter = false,
     }
 }
 
-local channelToDelete, wordToDelete = nil, nil -- used to store which respective table item to delete
+local channelToDelete, wordToDelete, filterwordsToDelete = nil, nil, nil -- used to store which respective table item to delete
```
Comment (brodyreeves) [inline config.lua]: Unnecessary due to being contained in 60d8731

File: config.lua [hunk 11]
```diff
@@ -8,11 +8,13 @@ MyChatAlert.defaults = {
         printOn = true,
         channels = {},
         words = {},
+        filterwords = {},
+        dedupTime = 0,
```
Comment (brodyreeves) [inline config.lua:12]: Useful for time filtering

File: config.lua [hunk 12]
```diff
@@ -75,6 +77,20 @@ MyChatAlert.options = {
...
+                dedup = {
+                    name = L["Ignore duplicated message in"],
+                    desc = L["unit second, 0 to disable"],
+                    type = "input", order = 2,
+                    get = function(info) return "" .. MyChatAlert.db.profile.dedupTime end,
+                    set = function(info, val) if val and val ~= "" then MyChatAlert.db.profile.dedupTime = tonumber(val) or 0 end end,
+                    disabled = function() return not MyChatAlert.db.profile.enabled end,
+                },
+            },
+        },
```
Comment (brodyreeves) [inline config.lua:93]: Useful for time filtering

File: config.lua [hunk 13]
```diff
@@ -163,9 +179,43 @@ MyChatAlert.options = {
...
+                        end
+                    end,
+                    disabled = function() return not MyChatAlert.db.profile.enabled end,
+                },
+            },
+        },
         miscOptions = {
             name = L["Misc Options"],
-            type = "group", inline = true, order = 7,
+            type = "group", inline = true, order = 8,
```
Comment (brodyreeves) [inline config.lua]: Unnecessary due to being contained in 60d8731

File: core.lua [hunk 14]
```diff
@@ -21,6 +21,8 @@ function MyChatAlert:OnDisable()
     self:UnregisterEvent("CHAT_MSG_CHANNEL")
 end
 
+local dedupmsg = {}
```
Comment (brodyreeves) [inline core.lua]: Useful for time filtering

File: core.lua [hunk 15]
```diff
@@ -36,17 +38,41 @@ function MyChatAlert:CHAT_MSG_CHANNEL(event, message, author, _, channel)
...
+            end
+
+            if not filtered then
+                for k2, word in pairs(self.db.profile.words) do
+                    if message:lower():find(word:lower()) then -- Alert message
+                        if self.db.profile.soundOn then PlaySound(self.db.profile.sound) end
+                        if self.db.profile.printOn then
+                            LibStub("AceConsole-3.0"):Print(format(L["Printed alert"], word, "|Hplayer:" .. TrimRealmName(author) .. ":0|h" .. TrimRealmName(author) .. "|h", message))
+                        end
+                        self:AddAlert(word, author, message)
```
Comment (brodyreeves) [inline core.lua]: Unnecessary due to being contained in 60d8731

File: core.lua [hunk 16]
```diff
@@ -36,17 +38,41 @@ function MyChatAlert:CHAT_MSG_CHANNEL(event, message, author, _, channel)
...
+                            dedupmsg[message] = time()
+                        end
+                        break -- matched the message so stop looping
                     end
-                    self:AddAlert(word, author, message)
-                    break -- matched the message so stop looping
                 end
+                break -- matched the channel so stop looping
             end
-            break -- matched the channel so stop looping
```
Comment (brodyreeves) [inline core.lua]: Unnecessary due to being contained in 60d8731

File: core.lua [hunk 17]
```diff
@@ -36,17 +38,41 @@ function MyChatAlert:CHAT_MSG_CHANNEL(event, message, author, _, channel)
...
+                for k2, word in pairs(self.db.profile.words) do
+                    if message:lower():find(word:lower()) then -- Alert message
+                        if self.db.profile.soundOn then PlaySound(self.db.profile.sound) end
+                        if self.db.profile.printOn then
+                            LibStub("AceConsole-3.0"):Print(format(L["Printed alert"], word, "|Hplayer:" .. TrimRealmName(author) .. ":0|h" .. TrimRealmName(author) .. "|h", message))
+                        end
+                        self:AddAlert(word, author, message)
+                        if self.db.profile.dedupTime > 0 then
+                            dedupmsg[message] = time()
+                        end
```
Comment (brodyreeves) [inline core.lua]: Useful for time filtering

**Review by brodyreeves — COMMENTED** [review]
My review comments are mostly for myself, so I can help filter through the changes easier then come back around to integrate the code

**Comment by tg123** [comment]
👍  
dungeon too much this week

**Pull request merged**